opentelemetry-exporter-geneva/geneva-uploader/src/payload_encoder/lz4_chunked_compression.rs

```diff
@@ -35,48 +35,49 @@ use lz4_flex::block::{compress_into, get_maximum_output_size};
 ///   but care must be taken: LZ4 does not natively support in-place compression, and the compressed size may be larger than the input.
 ///   If single-buffer "in-place" compression is possible (e.g., with unsafe or buffer aliasing), document or implement it here.
 /// - Consider passing output buffer as mutable slice to avoid reallocation, and provide another method to return the max size of the output buffer.
+///
+///
+/// Compresses input data in 64 KiB chunks using LZ4.
+/// Each chunk is prefixed with a 4-byte (little-endian) length header.
+/// See module documentation for layout and decompression notes.
 #[allow(dead_code)]
 pub(crate) fn lz4_chunked_compression(
     input: &[u8],
 ) -> Result<Vec<u8>, lz4_flex::block::CompressError> {
-    const CHUNK_SIZE: usize = 64 * 1024;
-    let max_chunk_compressed = get_maximum_output_size(CHUNK_SIZE);
+    lz4_chunked_compression_custom::<{ 64 * 1024 }>(input)
+}
 
-    // Pre-allocate an output buffer large enough for the worst-case total output size:
-    // Each chunk may require up to get_maximum_output_size(CHUNK_SIZE) bytes for compressed data,
-    // plus 4 bytes for the length header per chunk.
-    let mut output =
-        Vec::with_capacity(input.len().div_ceil(CHUNK_SIZE) * (4 + max_chunk_compressed));
+/// Compresses input data in user-defined chunk size using LZ4.
+/// Each chunk is prefixed with a 4-byte (little-endian) length header.
+pub(crate) fn lz4_chunked_compression_custom<const CHUNK_SIZE: usize>(
+    input: &[u8],
+) -> Result<Vec<u8>, lz4_flex::block::CompressError> {
+    if input.len() == 0 {
+        return Ok(Vec::new());
+    }
+    // Special case: very small data (less than one chunk)
+    if input.len() <= CHUNK_SIZE {
+        let max_compressed = get_maximum_output_size(input.len());
+        let mut temp_buf = vec![0u8; max_compressed];
+        let compressed_size = compress_into(input, &mut temp_buf)?;
+        let mut output = Vec::with_capacity(4 + compressed_size);
+        output.extend_from_slice(&(compressed_size as u32).to_le_bytes());
+        output.extend_from_slice(&temp_buf[..compressed_size]);
+        return Ok(output);
+    }
+
+    let max_chunk_compressed = get_maximum_output_size(CHUNK_SIZE);
+    let num_chunks = input.len().div_ceil(CHUNK_SIZE);
+    let mut output = Vec::with_capacity(num_chunks * (4 + max_chunk_compressed));
+    let mut temp_buf = vec![0u8; max_chunk_compressed];
```

Comment:
I don't follow. We had removed all temporary buffer allocation. Why are we introducing it again?

Reply:
Good point. This is not correct. I see improvements in the small size with this, but not with large size. Have reverted the optimization efforts - the change is now limited to using the benchmark! Thanks for catching this.